### PR TITLE
Regex boundary clean and various tweaks to reduce FPs (part 1)

### PR DESCRIPTION
**Pull request by zricethezav**
### Description:
Adds better regex boundaries and includes various tweaks to reduce FPs

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/detectors/appointedd/appointedd.go [hunk 1]
```diff
@@ -21,7 +22,7 @@ var (
 	client = common.SaneHttpClient()
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives
-	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"appointedd"}) + `\b([a-zA-Z0-9=+]{88})`)
+	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"appointedd"}) + `\b([a-zA-Z0-9=+]{88})\b`)
```
Comment (rgmz) [inline pkg/detectors/appointedd/appointedd.go:25]: `\b` doesn't work unless the character range is `\w` (`[a-z0-9_`). So if this, or any other pattern, ends in `=` it won't match.

File: pkg/detectors/datadogtoken/datadogtoken.go [hunk 2]
```diff
@@ -20,8 +21,8 @@ var (
 	client = common.SaneHttpClient()
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
-	appPat = regexp.MustCompile(detectors.PrefixRegex([]string{"datadog", "dd"}) + `\b([a-zA-Z-0-9]{40})\b`)
-	apiPat = regexp.MustCompile(detectors.PrefixRegex([]string{"datadog", "dd"}) + `\b([a-zA-Z-0-9]{32})\b`)
+	appPat = regexp.MustCompile(detectors.PrefixRegex([]string{"datadog", "dd_", "dd-"}) + `\b([a-zA-Z-0-9]{40})\b`)
```
Comment (zricethezav) [inline pkg/detectors/datadogtoken/datadogtoken.go:24]: For the reviewer: I don't feel super strongly on this change. However it does seem like `dd` alone could introduce a fair number of FPs. Open for discussion!

File: pkg/detectors/dovico/dovico.go [hunk 3]
```diff
@@ -21,8 +22,8 @@ var (
 	client = common.SaneHttpClient()
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
-	keyPat  = regexp.MustCompile(detectors.PrefixRegex([]string{"dovico"}) + `\b([0-9a-z]{32}\.[0-9a-z]{1,}\b)`)
-	userPat = regexp.MustCompile(detectors.PrefixRegex([]string{"dovico"}) + `\b([0-9a-z]{32}\.[0-9a-z]{1,}\b)`)
+	keyPat  = regexp.MustCompile(detectors.PrefixRegex([]string{"dovico"}) + `\b([0-9a-z]{32}\.[0-9a-z]{1,100})\b`)
```
Comment (zricethezav) [inline pkg/detectors/dovico/dovico.go:25]: for the reviewer: _probably_ not gonna go past 100 alpanums on the backend here but 🤷🏻 I'm open for discussion

File: pkg/detectors/gengo/gengo.go [hunk 4]
```diff
@@ -29,8 +30,8 @@ var (
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
 	// Removed bounds since there are some cases where the start and end of the token is a special character
-	keyPat    = regexp.MustCompile(detectors.PrefixRegex([]string{"gengo"}) + `([ ]{0,1}[0-9a-zA-Z\[\]\-\(\)\{\}|_^@$=~]{64}[ \r\n]{1})`)
-	secretPat = regexp.MustCompile(detectors.PrefixRegex([]string{"gengo"}) + `([ ]{0,1}[0-9a-zA-Z\[\]\-\(\)\{\}|_^@$=~]{64}[ \r\n]{1})`)
+	keyPat    = regexp.MustCompile(detectors.PrefixRegex([]string{"gengo"}) + `([0-9a-zA-Z\[\]\-\(\)\{\}|_^@$=~]{64})\b`)
```
Comment (zricethezav) [inline pkg/detectors/gengo/gengo.go:33]: For the reviewer: this token has a real nasty pattern to match. No `\b` in front of the capture group due to the fact that some of the token's alphabet _could_ be escaped by that in-front-of-capture-group `\b`. `[ ]{0,1}` removed as it didn't provide any value according to my tests (though please double check this work if you can). 

FWIW the resulting regex is: `(?i)(?:gengo)(?:.|[\n\r]){0,40}([0-9a-zA-Z\[\]\-\(\)\{\}|_^@$=~]{64})\b`

We should probably have a discussion on the fact that the keyPat and secretPat are the same regex. I'm sure there is a way to tighten something up here.

File: pkg/detectors/appointedd/appointedd.go (hunk 1 above)
Comment (zricethezav) [reply pkg/detectors/appointedd/appointedd.go]: There are quite a few regexes like this. I'm wondering if we should use a more strict regex like the gitleaks prefix:
```
(?i)(?:provider)(?:[0-9a-z\-_\t .]{0,20})(?:[\s|']|[\s|"]){0,3}(?:=|>|:{1,3}=|\|\|:|<=|=>|:|\?=)(?:'|\"|\s|=|\x60){0,5}
```

File: pkg/detectors/aiven/aiven.go [hunk 5]
```diff
@@ -21,7 +22,7 @@ var (
 	client = common.SaneHttpClient()
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
-	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"aiven"}) + `([a-zA-Z0-9/+=]{372})`)
+	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"aiven"}) + `\b([a-zA-Z0-9/+=]{372})\b`)
```
Comment (zricethezav) [inline pkg/detectors/aiven/aiven.go:25]: The previous regex fails detecting our test secret as does the new one. This one needs more work still

File: pkg/detectors/gengo/gengo.go (hunk 4 above)
Comment (zricethezav) [reply pkg/detectors/gengo/gengo.go:33]: Same issue as @rgmz pointed out in https://github.com/trufflesecurity/trufflehog/pull/2407#discussion_r1484813476

File: pkg/detectors/appointedd/appointedd.go (hunk 1 above)
Comment (rgmz) [reply pkg/detectors/appointedd/appointedd.go]: I think we'd need more data to determine if that'd be worthwhile. 

Another option would be to do something like this, which IIRC is what GitHub does.
```
[a-zA-Z0-9=+]{88}(?:[^a-zA-Z0-9=+]|\z)
```

File: pkg/detectors/appointedd/appointedd.go [hunk 6]
```diff
@@ -21,7 +22,7 @@ var (
 	client = common.SaneHttpClient()
 
 	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives
-	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"appointedd"}) + `\b([a-zA-Z0-9=+]{88})`)
+	keyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"appointedd"}) + `\b([a-zA-Z0-9=+]{87}=)`)
```
Comment (dustin-decker) [inline pkg/detectors/appointedd/appointedd.go:25]: Might not always have base64 padding?

Comment (zricethezav) [reply pkg/detectors/appointedd/appointedd.go:25]: I think it always is base64 padded. I just created a new one and it's b64. The test secret we have on deck is also b64. Was there another case you were think of?

File: pkg/detectors/appointedd/appointedd.go (hunk 1 above)
Comment (zricethezav) [reply pkg/detectors/appointedd/appointedd.go]: Assuming appointed is always b64 encoded, which I think it is, the boundary could just be `=` as it's proposed in the current change

https://regex101.com/r/YJeAfP/1

File: pkg/detectors/aiven/aiven.go (hunk 5 above)
Comment (bugbaba) [reply pkg/detectors/aiven/aiven.go:25]: Just update the number of chars to 372 to 380, Also boundary won't work for this as it contains special chars within the match/key.

Comment (zricethezav) [reply pkg/detectors/aiven/aiven.go:25]: > Just update the number of chars from 372 to 380

I'm not following how this would solve the problem.

Comment (zricethezav) [reply pkg/detectors/aiven/aiven.go:25]: Generated a couple tokens just now and they all have `377` characters

Comment (zricethezav) [reply pkg/detectors/aiven/aiven.go:25]: alas, the problem still persists https://regex101.com/r/nMLU8f/1. Possible solution: https://github.com/trufflesecurity/trufflehog/pull/2407#discussion_r1484822724

Comment (rgmz) [reply pkg/detectors/aiven/aiven.go:25]: This is dumb, but try adding a `?` before the `\b`. That works in your linked example at least.

Comment (bugbaba) [reply pkg/detectors/aiven/aiven.go:25]: Super weird last month I had tried multiple times it was generating  keys of `380` length each time. Will try again.

Comment (bugbaba) [reply pkg/detectors/aiven/aiven.go:25]: Just confirmed it, Its still generating keys of `380` length for me. 
Just to be sure you are creating tokens on https://console.aiven.io/profile/tokens page right?

https://regex101.com/r/MSXILB/1

Comment (zricethezav) [reply pkg/detectors/aiven/aiven.go:25]: >  Just to be sure you are creating tokens on https://console.aiven.io/profile/tokens page right?

Yep. Try creating an account with another email and generating some tokens

Comment (bugbaba) [reply pkg/detectors/aiven/aiven.go:25]: Super Weird, Tried with new account and now its generating token of `372` length

Comment (rgmz) [reply pkg/detectors/aiven/aiven.go:25]: It's probably encoding account-specific data, which may explain the variance.

Comment (zricethezav) [reply pkg/detectors/aiven/aiven.go:25]: > This is dumb, but try adding a ? before the \b. That works in your linked example at least.

Lazy loading does the trick. Good eye @rgmz

File: pkg/detectors/appointedd/appointedd.go (hunk 6 above)
Comment (zricethezav) [reply pkg/detectors/appointedd/appointedd.go]: reverting back to the original

**Comment by zricethezav** [comment]
Going to close this PR and start fresh. After doing some testing some of these `\b`s are misplaced and have negative outcomes

**Pull request closed without merging**